### PR TITLE
fixes for #13767

### LINKED ISSUE #13767: Make (TLS) groups fully independent of (EC) curves
Referring to the statement by @mattcaswell in #10512

> We have implemented fully pluggable KEM and KEX in libssl

wouldn't the expectation be fair that after adding new TLS (1.3) groups (via a pluggable KEM provider) those new groups should be visible/accessible in a similar way as the original (EC) curves? 

More specifically, the API `SSL_get_shared_group` seems to behave somewhat out of line with this full integration: By returning only NIDs (of known libcrypto implementations), it effectively returns nothing if a [new, Provider-implemented, TLS1.3 KEM group (PQC in this case)](https://github.com/open-quantum-safe/openssl/tree/OQS-OpenSSL3) has been integrated and runs:

```
Supported Elliptic Groups: 0x0201
Shared Elliptic groups: <NULL>
```

TLS1.3 operates correctly with the new crypto (KEM) from the Provider (confirming the statement at the beginning), but only tries to display built-in/libcrypto-provided shared (Elliptic) groups.

In any case, the output above seems to be inconsistent: Either it should display nothing (as the new KEM is not an EC curve) OR also display the shared group (which already is shown as supported with its group ID). The reason for the NULL above is this statement in `ssl/s3_lib.c`:

```
return tls1_group_id2nid(id, 1);
```

A possible alternative is to amend the above return statement as follows

```
if (s->session->ssl_version != TLS1_3_VERSION) return tls1_group_id2nid(id, 1);
```

and update the documentation in `doc/man3/SSL_CTX_set1_curves.pod` accordingly: Amend the current

> SSL_get_shared_group() returns the NID of the shared group B<n> for a server-side SSL B<ssl>.

as follows:

> SSL_get_shared_group() returns the NID of the shared group B<n> for a server-side SSL B<ssl> when running SSL or TLS<1.3; for TLS1.3, returns the shared group ID.

In addition, it would be nice to amend the code in `apps/lib/s_cb.c` annotated "TODO(TLS1.3)" to fetch the group name. When trying to implement this I had to find I'm out of my depth here: At first look, the function `tls1_group_id_lookup` could deliver the `TLS_GROUP_INFO` structure containing all required information, however, function and data structures seem to be only (SSL-)local and not generally accessible, so some internal "API opening" may be be required for this. So, any reference to APIs I am not aware of and already delivering this functionality would be gratefully welcomed: I'd volunteer to integrate that into `apps/lib/s_cb.c`.

An alternative to this proposal would be a new API for TLS1.3 groups that has no relation to the old (EC) curves.

### PR DESCRIPTION
**Pull request by baentsch**
Fixes #13767:
- Implements TLS1.3 ToDos (group name retrieval) and fixes incorrect EC output reference in `apps/lib/s_cb.c`
- Eliminates incorrect OPENSSL_NO_EC and OPENSSL_NO_DH code guards for group-accessor functions and data structures

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
There is no real commit message. The Fixes #... should be on a separate line in commit message bod, the commit header line should summarize the change with commit message body giving more details.

Also please unless absolutely necessary you should use a real e-mail address for commits - and it needs to be the address that you submitted with the ICLA. Otherwise the cla-check will fail.

**Comment by mattcaswell** [comment]
Close/reopen to kick CLA bot

**Comment by mattcaswell** [comment]
I added the anonymous github email address as an alias so that the CLA check passes.

**Comment by baentsch** [comment]
> There is no real commit message. The Fixes #... should be on a separate line in commit message body, the commit header line should summarize the change with commit message body giving more details.
> 
> Also please unless absolutely necessary you should use a real e-mail address for commits - and it needs to be the address that you submitted with the ICLA. Otherwise the cla-check will fail.

Thanks for the guidance. Looks like I'm already relying too much on the git UI and less on its command line features. Will remember the next time. I also changed user.email to a less anonymous one now.

**Comment by t8m** [comment]
To fix the commit message you can use `git commit --amend` command. Then you just have to do `git push --force`.

**Comment by baentsch** [comment]
> To fix the commit message you can use `git commit --amend` command. Then you just have to do `git push --force`.

Thanks. So done.

**Comment by romen** [comment]
I agree with @kaduk comment about separating the different (related) concerns in different commits: you could edit your git history with an interactive rebase and then force-push to this same branch if you are willing to do so.

This way the history of feedback given in this PR is not lost! 

Also, thanks again for working on this issue!

**Comment by t8m** [comment]
@romen Your comment seems misplaced from another PR? I do not even see @kaduk commenting here.

File: apps/lib/s_cb.c
Comment (t8m) [inline apps/lib/s_cb.c:365]: Should be `||` instead?

File: apps/lib/s_cb.c
Comment (t8m) [inline apps/lib/s_cb.c:385]: Same as above.

**Comment by romen** [comment]
There was discussion in the related issue: https://github.com/openssl/openssl/issues/13767#issuecomment-755114640

File: test/sslapitest.c
Comment (t8m) [inline test/sslapitest.c:8003]: Please wrap the line.

File: include/openssl/ssl.h.in
Comment (t8m) [inline include/openssl/ssl.h.in:1504]: Please add space after `char`

File: ssl/s3_lib.c
Comment (t8m) [inline ssl/s3_lib.c:4975]: Please add space after char

File: doc/man3/SSL_groupid_to_name.pod
Comment (t8m) [inline doc/man3/SSL_groupid_to_name.pod:15]: This is my biggest concern about this PR. This returns the algorithm name which is actually not that interesting by itself. We want the real group name to be presented. I think either the tlsname or realname of the group would be more useful.

File: doc/man3/SSL_groupid_to_name.pod
Comment (t8m) [inline doc/man3/SSL_groupid_to_name.pod:27]: `I<ssl>`

File: doc/man3/SSL_groupid_to_name.pod
Comment (t8m) [inline doc/man3/SSL_groupid_to_name.pod:16]: `I<group_id>`
However it would be also better to say, what the group_id number actually is. That it is the TLS group ID.

File: doc/man3/SSL_groupid_to_name.pod
Comment (t8m) [inline doc/man3/SSL_groupid_to_name.pod:21]: Please no `B<>` around `NULL`.

**Review by t8m — CHANGES_REQUESTED** [review]
Good work. I have a few nits and one concern about the SSL_groupid_to_name definition.

File: apps/lib/s_cb.c
Comment (baentsch) [reply apps/lib/s_cb.c:365]: Nope: If we have a name we want to show that (instead of the group ID).

File: test/sslapitest.c
Comment (baentsch) [reply test/sslapitest.c:8003]: Oops - will do - clearly didn't read https://www.openssl.org/policies/codingstyle.html in time :-(

File: apps/lib/s_cb.c
Comment (t8m) [reply apps/lib/s_cb.c:365]: Then it should simply be `if (gname == NULL)`.

**Comment by baentsch** [comment]
> you could edit your git history with an interactive rebase and then force-push to this same branch if you are willing to do so.

Well, then I must admit I'm apparently not sufficiently educated in rebasing: I thought one can use rebasing only to "fiddle with" separate commits if they already exist, but not retroactively create new "piece-part" commits after having done everything in one go. Or is the idea to `git reset --hard` the branch, add piece-part commits until we arrive at the same state again (heeding the feedback above), and then `push --force again`? Or do I misunderstand you @romen?

**Comment by romen** [comment]
Yes, it is basically the same thing with just one commit, I guess I am mostly using the interactive rebase as that way it is easier to retain commit messages and to abort and go back to original state if I make some mistake.
You can use whatever strategy you are more confident with!

File: apps/lib/s_cb.c
Comment (baentsch) [reply apps/lib/s_cb.c:365]: That would then change the output (i.e., externally visible behaviour) of the function compared to before: If the returned NID does not have the `TLSEXT_nid_unknown` bit set (and no name has been found), it would simply (unbeknownst to the user) output the NID instead of the group_id. Until now, this case became visible by receiving the output `(nil)`. I don't know whether this can happen (having no group name _and_ `TLSEXT_nid_unknown` unset), but the additional guard made sense to me from a safety perspective. But I'd surely follow your judgement in this regard, so will change if I don't get additional feedback.

Comment (t8m) [reply apps/lib/s_cb.c:365]: There is simply no reason to output nil in such case. It does not provide any useful information.

Comment (baentsch) [reply apps/lib/s_cb.c:365]: It only helped me pinpoint quickly missing support for non-built-in groups; but as I'm all for changing something that wasn't sensible, `if (gname == NULL)` it will be.

File: apps/lib/s_cb.c
Comment (mattcaswell) [inline apps/lib/s_cb.c:363]: This step should not be necessary. If we its a supported group, then we should know the name. A `gname` of NULL would indicate some error in that case.

File: apps/lib/s_cb.c
Comment (mattcaswell) [reply apps/lib/s_cb.c:365]: `SSL_groupid_to_name` should work and give a sensible name for all outputs from `SSL_get1_groups` and similar functions. Therefore it should be able to handle just a nid, or a group id with the `TLSEXT_nid_unknown` bit set. So I don't think this code printing out the hex is necessary. Either `gname` is non-null in which case we print it, or alternatively it is NULL in which case an error has occurred.

File: doc/man3/SSL_groupid_to_name.pod
Comment (mattcaswell) [reply doc/man3/SSL_groupid_to_name.pod:15]: It should always be the tlsname IMO.

File: doc/man3/SSL_groupid_to_name.pod
Comment (mattcaswell) [reply doc/man3/SSL_groupid_to_name.pod:16]: I don't think that is the correct semantics for this function. It should accept the nid, or the group id with the `TLSEXT_nid_unknown` bit set - but not the plain group id. That way it can be used to print the name on the results of functions like `SSL_get_shared_group`. For this reason I would call the function `SSL_group_to_name()` and not `SSL_groupid_to_name()`.

File: doc/man3/SSL_groupid_to_name.pod
Comment (mattcaswell) [inline doc/man3/SSL_groupid_to_name.pod:35]: Just 2021

File: include/openssl/ssl.h.in
Comment (mattcaswell) [reply include/openssl/ssl.h.in:1504]: And remove the space after `*`, i..e.

````
const char *SSL_groupid_to_name(SSL *s, int group_id);
````

File: ssl/s3_lib.c
Comment (mattcaswell) [inline ssl/s3_lib.c:4976]: This works by accident. `tls1_group_id_lookup` takes a `uint16_t` as its second parameter. Here you are doing an implicit cast from `int` to `uint16_t`. That means if the input value happens to have the `TLSEXT_nid_unknown` bit set then that bit gets dropped and it works. This is why your test works. But it won't work for real nids - which it should do.

I think we should introduce a new internal function `tls1_nid2group` which works a bit like `tls1_nid2group_id` but returns a `TLS_GROUP_INFO *` and is able to handle the `TLSEXT_nid_unknown` semantics.

File: test/sslapitest.c
Comment (mattcaswell) [reply test/sslapitest.c:8003]: As noted above this only works (by accident) if `SSL_get_shared_group` happens to return a group with the `TLSEXT_nid_unknown` bit set. Otherwise it will return NULL.

I think we should have another test of this function that checks that a "built-in" group does the right thing.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: apps/lib/s_cb.c
Comment (baentsch) [reply apps/lib/s_cb.c]: ACK. Test removed. Error (NULL) will become printed.

File: apps/lib/s_cb.c
Comment (baentsch) [reply apps/lib/s_cb.c]: see above: Very much simplified.

File: test/sslapitest.c
Comment (baentsch) [reply test/sslapitest.c]: So added (a set of them, actually).

File: include/openssl/ssl.h.in
Comment (baentsch) [reply include/openssl/ssl.h.in]: Done.

File: ssl/s3_lib.c
Comment (baentsch) [reply ssl/s3_lib.c]: Done.

File: doc/man3/SSL_groupid_to_name.pod
Comment (baentsch) [reply doc/man3/SSL_groupid_to_name.pod]: Done.

File: doc/man3/SSL_groupid_to_name.pod
Comment (baentsch) [reply doc/man3/SSL_groupid_to_name.pod]: Done.

File: doc/man3/SSL_groupid_to_name.pod
Comment (baentsch) [reply doc/man3/SSL_groupid_to_name.pod]: Done.

File: doc/man3/SSL_groupid_to_name.pod
Comment (baentsch) [reply doc/man3/SSL_groupid_to_name.pod]: Done.

File: apps/lib/s_cb.c
Comment (baentsch) [reply apps/lib/s_cb.c]: Agree.

File: doc/man3/SSL_groupid_to_name.pod
Comment (baentsch) [reply doc/man3/SSL_groupid_to_name.pod]: Done.

File: ssl/s3_lib.c
Comment (baentsch) [reply ssl/s3_lib.c]: Agreed. Please check new code as to whether it correctly captures the above. I refrained from adding another function as `tls1_nid2group_id` could be utilized suitably.

Comment (baentsch) [reply ssl/s3_lib.c]: Addressed your correct comments with the update - in a slightly different way (not adding a new function, but using an existing one). 

A final "etiquette" question: Please let me know whether I should click "Resolve conversation" if I think I followed through or whether the person raising the objection will do that after reviewing my comments/code changes.

File: ssl/s3_lib.c
Comment (mattcaswell) [inline ssl/s3_lib.c:5006]: Drop the "else" and just make this "return NULL", since we've already returned in the other branch.

File: test/sslapitest.c
Comment (mattcaswell) [inline test/sslapitest.c:4443]: `TEST_str_eq` is probably better here

File: test/sslapitest.c
Comment (mattcaswell) [inline test/sslapitest.c:8022]: And here

**Review by mattcaswell — CHANGES_REQUESTED** [review]
LGTM. Just a few minor nits remaining.

File: ssl/s3_lib.c
Comment (mattcaswell) [reply ssl/s3_lib.c]: Feel free to click "resolve" if you believe that you've fully fixed an issue. Rest assured that a reviewer will "unresolve" if they disagree :-)

Comment (romen) [reply ssl/s3_lib.c]: > A final "etiquette" question: Please let me know whether I should click "Resolve conversation" if I think I followed through or whether the person raising the objection will do that after reviewing my comments/code changes.

I don't think we have a codified practice on this, my personal rule of thumb is that, as the PR author, I tend to click "Resolve conversation" myself only in the trivial ones (e.g. "there is a missing space here", "typo there") after writing a "Done/Fixed/Addressed" comment.
Instead, I leave it to whomever started the discussion for the other cases, especially if I'd like to have some explicit confirmation that my fix fully addressed the issue as intended by the reviewer.

But again, this is just my rule of thumb.

File: test/sslapitest.c
Comment (baentsch) [reply test/sslapitest.c]: done

File: ssl/s3_lib.c
Comment (baentsch) [reply ssl/s3_lib.c]: Done

File: test/sslapitest.c
Comment (baentsch) [reply test/sslapitest.c]: Thanks for the pointer to that function! So now used - again better code.

File: test/sslapitest.c
Comment (mattcaswell) [inline test/sslapitest.c:8023]: Indentation looks a little odd here

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Approved subject to the nit below being fixed. Needs a second review.

File: test/sslapitest.c
Comment (t8m) [reply test/sslapitest.c:8023]: Yeah, better split the line after the group_name. However this would mean with proper indent that the second line will be still slightly longer than 80 chars. But I think that should be acceptable.

Comment (baentsch) [reply test/sslapitest.c]: Done -- indeed slightly above the 80... but if you'd think that's OK, indentation certainly is more logical now.

**Review by mattcaswell — APPROVED** [review]

**Review by romen — APPROVED** [review]
LGTM, thanks!

**Review by t8m — APPROVED** [review]
LGTM now. I'll mark as approval done and leave it up to @mattcaswell to clear it if he thinks the indentation/line split needs to be still improved.

**Comment by romen** [comment]
Also, whoever merges this, please remember to squash the commits together!

@baentsch one nice way I learned about while working on openssl, to keep the "addressing feedback" commits separated but still have the commits squashed when we will merge is: `git commit --fixup=<commit_ref>` (you can also add `-e` if you still want to edit the commit message to include some notes, but they are going o be discarded when rebasing with `git rebase --autosquash` upon merge).

**Comment by mattcaswell** [comment]
> LGTM now. I'll mark as approval done and leave it up to @mattcaswell to clear it if he thinks the indentation/line split needs to be still improved.

My approval stands

**Comment by mattcaswell** [comment]
Pushed! Thanks for your contribution.

**Comment by kaduk** [comment]
> Also, whoever merges this, please remember to squash the commits together!
> 
> @baentsch one nice way I learned about while working on openssl, to keep the "addressing feedback" commits separated but still have the commits squashed when we will merge is: `git commit --fixup=<commit_ref>` (you can also add `-e` if you still want to edit the commit message to include some notes, but they are going o be discarded when rebasing with `git rebase --autosquash` upon merge).

@romen  There's also `git commit --squash=<commit_ref>` that lets you write commit messages that will not be combined, not discarded, with `git rebase --autosquash`.

**Pull request closed without merging**
